In [2]:
import pandas as pd
import numpy as np
import scipy.linalg as sla

%matplotlib inline
import matplotlib.pyplot as plt

#### Exercicio 1

In [7]:
def elimin_gaussiana(a,b,imp = False, norma='inf'):
     #matriz estendida Me
    b = np.mat(b).T
    Me = np.concatenate((a,b), axis=1)
    if imp: print('Matriz estendida: \n',Me,'\n\n')
    if imp: print('='*80)
    for j in range(Me.shape[-1] - 2): #coluna
        if imp: print('Operação na coluna', j)
        if imp: print('='*80, '\n\n')
        for i in range(j+1, Me.shape[0]): #linha
            if (Me[j,j] != 0):
                Me[i] = Me[i] - (Me[i,j]/Me[j,j])*Me[j] #operações com cada linha
            else:
                Me[i] = Me[i] - (0)*Me[j]
            if imp: print('Operação na linha', i, ':',Me[i], '\n')
            if imp: print('Matriz escalonando: \n',Me,'\n')
        if imp: print('='*80)
    if imp: print('\nMatriz escalonada: \n',Me,'\n')
    Matriz_A = np.delete(Me, -1, axis=1) #matriz de coeficientes escalonada
    Vetor_b = Me[:,-1] #matriz de termos independentes escalonada
    Vetor_x = sla.solve_triangular(Matriz_A, Vetor_b) # solução do sistema escalonado
    if imp: print('='*80,'\n\nSolução do sistema: \n', Vetor_x)
    #verificação do erro do sistema
    if norma is 'inf': norma = np.inf
    if norma is '-inf': norma = -np.inf
    Erro = sla.norm(b - a.dot(Vetor_x),ord = norma)
    return (Matriz_A, Vetor_b, Vetor_x, Erro)

In [11]:
a = np.array([[2,3,-1],[4,4,-3],[2,-3,1]])
b = np.array([5,3,-1])
M = elimin_gaussiana(a,b)
M

(matrix([[ 2,  3, -1],
         [ 0, -2, -1],
         [ 0,  0,  5]]), matrix([[ 5],
         [-7],
         [15]]), array([[1.],
        [2.],
        [3.]]), 0.0)

#### Exercicio 2

O erro pode ser medido por $||ax-b|| = erro$, uma vez que um sistema linear é sempre da seguinte forma $ax = b$, onde havendo solução (determinada ou não) $ax-b = 0$. O erro em questão ocorre pois o pivô passa a ser 0(causando uma divisão por 0).

#### Exercicio 3

In [22]:
A = np.array([[1,4,52],[27,110,-3],[22,2,14]])
b = np.array([57,134,38])
sla.solve(A,b)

array([1., 1., 1.])

#### Exercicio 4

In [8]:
def  elim_gaussiana2(A, b, pivotamento=None, imp=False, norma='inf'):
    b = np.mat(b).T
    me = np.array(np.concatenate((A, b), axis=1))
    n = len(me)

    if imp:
        print('Matriz estendida:\n{}\n'.format(me))
    if imp:
        print('=' * 80)

    # Eliminação
    for k in range(n - 1):
        if pivotamento:
            # Pivô
            maxindex = abs(me[k:, k]).argmax() + k
            if me[maxindex, k] == 0:
                raise ValueError("Matriz é singular.")
            # Trocas
            if maxindex != k:
                me[[k, maxindex]] = me[[maxindex, k]]

                if imp:
                    print("Troca entre as linhas {} e {}".format(k, maxindex))
                if imp:
                    print('Matriz escalonando:\n{}\n'.format(me))
        else:
            if me[k, k] == 0:
                raise ValueError("Pivô é zero. Tente com a variável \'pivoteamento\' como True.")

        for linha in range(k + 1, n):
            mult = me[linha, k] / me[k, k]
            me[linha, k:] = me[linha, k:] - mult * me[k, k:]
            if imp:
                print('Operação na linha {}: {}'.format(linha, me[linha]))
            if imp:
                print('Matriz escalonando: \n {} \n'.format(me))

    # Retrosubstituição

    if imp:
        print('Matriz escalonada:\n{}\n'.format(me))

    matriz_a = me[:, :-1]  # matriz de coeficientes escalonada
    vetor_b = me[:, -1]  # matriz de termos independentes escalonada
    vetor_x = sla.solve_triangular(matriz_a, vetor_b)  # solução do sistema escalonado

    if imp:
        print('=' * 80, '\n\nSolução do sistema: \n', vetor_x)

    # Verificação do erro do sistema
    if norma is 'inf':
        norma = np.inf
    if norma is '-inf':
        norma = -np.inf
    erro = sla.norm(b - A.dot(vetor_x), ord=norma)

    return matriz_a, vetor_b, vetor_x, erro

In [23]:
A= np.array([[1., 4, 52], [27, 110, -3], [22, 2, 14]])
b = np.array([57., 134, 38])
elim_gaussiana2(A, b, pivotamento=True)

(array([[ 27.        , 110.        ,  -3.        ],
        [  0.        , -87.62962963,  16.44444444],
        [  0.        ,   0.        ,  52.09721048]]),
 array([134.        , -71.18518519,  52.09721048]),
 array([1., 1., 1.]),
 173.0)

#### Exercicio 5

In [31]:
e = 10
A = np.array([[e,2],[1,e]])
b = np.array([4,3])
eList = []

resultPivotiamento = []
resultSemPivotiamento = []
for i in range(100):
    resultPivotiamento.append(elim_gaussiana2(A, b, pivotamento=True)[2])
    resultSemPivotiamento.append(elim_gaussiana2(A, b, pivotamento=False)[2])
    eList.append(e)
    e = 10**(-i)
    A= np.array([[e,2],[1,e]])
pd.DataFrame({'Pivoteamento=True':resultPivotiamento,'Pivoteamento=False':resultSemPivotiamento, "e":eList})

Pivoteamento=True  \
0   [0.3555555555555555, 0.2222222222222222]   
1                                 [2.0, 1.0]   
2    [2.814070351758794, 1.8592964824120604]   
3   [2.9801490074503727, 1.9850992549627482]   
4   [2.9980014990007495, 1.9985009992504994]   
5     [2.9998000149990003, 1.99985000999925]   
6    [2.999980000149999, 1.9999850000999992]   
7          [2.9999980000015, 1.999998500001]   
8      [2.999999800000015, 1.99999985000001]   
9                  [2.99999998, 1.999999985]   
10               [2.999999998, 1.9999999985]   
11             [2.9999999998, 1.99999999985]   
12           [2.99999999998, 1.999999999985]   
13         [2.999999999998, 1.9999999999985]   
14       [2.9999999999998, 1.99999999999985]   
15     [2.99999999999998, 1.999999999999985]   
16   [2.999999999999998, 1.9999999999999984]   
17                 [3.0, 1.9999999999999998]   
18                                [3.0, 2.0]   
19                                [3.0, 2.0]   
20                                [3.0, 2.0]   
21                                [3.0, 2.0]   
22                                [3.0, 2.0]   
23                                [3.0, 2.0]   
24                                [3.0, 2.0]   
25                                [3.0, 2.0]   
26                                [3.0, 2.0]   
27                                [3.0, 2.0]   
28                                [3.0, 2.0]   
29                                [3.0, 2.0]   
..                                       ...   
70                                [3.0, 2.0]   
71                                [3.0, 2.0]   
72                                [3.0, 2.0]   
73                                [3.0, 2.0]   
74                                [3.0, 2.0]   
75                                [3.0, 2.0]   
76                                [3.0, 2.0]   
77                                [3.0, 2.0]   
78                                [3.0, 2.0]   
79                                [3.0, 2.0]   
80                                [3.0, 2.0]   
81                                [3.0, 2.0]   
82                                [3.0, 2.0]   
83                                [3.0, 2.0]   
84                                [3.0, 2.0]   
85                                [3.0, 2.0]   
86                                [3.0, 2.0]   
87                                [3.0, 2.0]   
88                                [3.0, 2.0]   
89                                [3.0, 2.0]   
90                                [3.0, 2.0]   
91                                [3.0, 2.0]   
92                                [3.0, 2.0]   
93                                [3.0, 2.0]   
94                                [3.0, 2.0]   
95                                [3.0, 2.0]   
96                                [3.0, 2.0]   
97                                [3.0, 2.0]   
98                                [3.0, 2.0]   
99                                [3.0, 2.0]   

                          Pivoteamento=False             e  
0   [0.3555555555555555, 0.2222222222222222]  1.000000e+01  
1                                 [2.0, 1.0]  1.000000e+00  
2    [2.814070351758793, 1.8592964824120604]  1.000000e-01  
3    [2.9801490074504056, 1.985099254962748]  1.000000e-02  
4   [2.9980014990007575, 1.9985009992504996]  1.000000e-03  
5   [2.9998000149955573, 1.9998500099992502]  1.000000e-04  
6    [2.999980000151225, 1.9999850000999992]  1.000000e-05  
7       [2.9999979997974435, 1.999998500001]  1.000000e-06  
8      [2.999999799691011, 1.99999985000001]  1.000000e-07  
9           [2.999999981767587, 1.999999985]  1.000000e-08  
10         [2.999999804131903, 1.9999999985]  1.000000e-09  
11        [3.000000248221113, 1.99999999985]  1.000000e-10  
12       [3.000000248221113, 1.999999999985]  1.000000e-11  
13      [2.999822612537173, 1.9999999999985]  1.000000e-12  
14    [3.0020430585864233, 1.99999999999985]  1.000000e-13  
15    [3.019806626980426, 1.999999999999985]  1.000000e-14  
16   [3.10862

#### Exercicio 6

Os resultados são diferentes pois o método lu retorna as matrizes L e U separadas, permitindo outras aplicações desse resultado sem a necessidade de manipulações nas matrizes.

Já o método lu_factor() retorna ambas as matrizes L e U unidas em uma só. Ou seja, no formato ideal para ser usado como entrada para a função lu_solve.

#### Exercicio 7

Transformando os dados em matrizes e aplicando o algoritmo de solução.

In [10]:
matA = np.array([[225,0,-25,0], [0, 175, 0, -125], [225, 0, -275, 50], [0, 25, 250, -275]])
matB = np.array([1400,100, -2000, 0])
res = elim_gaussiana2(matA, matB, pivotamento=True)[2]
res

array([ 8.09983897, 12.35196687, 16.89855072, 16.49275362])

##### b) 
Para saber a quantidade de cada fator na área das crianças, devemos ter a matriz $b$ contendo apenas os valores correspondentes ao peso desse fator. Ex: para os fumantes, temos: $b = [1000,0,0,0 ]$

In [11]:
matB = np.array([1000,0,0,0])
res1 = elim_gaussiana2(matA, matB, pivotamento=True)[2]
res1

array([4.99624262, 3.45065562, 4.96618357, 4.83091787])

##### 1
Como nosso interesse é apenas na sala 2, temos que dos 12.35 mg/h de monóxido de carbono que chegam a sala 2, 3.45 são devido aos fumantes, correspondendo à cerca de 27.94% da poluição na sala das crianças.

In [13]:
matB = np.array([0,0,-2000,0])
res2 = elim_gaussiana2(matA, matB, pivotamento=True)[2]
res2

array([1.10359635, 6.90131125, 9.93236715, 9.66183575])

##### 2
A grelha corresponde a 58.77% da poluição da sala das crianças.

In [15]:
matB = np.array([400,100,0,0])
res3 = elim_gaussiana2(matA, matB, pivotamento=True)[2]
res3

array([2., 2., 2., 2.])

##### 3
A ventilação Qa é de aproximadamente 12.13% da poluição na sala das crianças

##### c)

In [16]:
A = np.array([[225, 0, -25, 0], [0, 175, 0, -125], [225, 0, -275, 50], [0, 25, 250, -275]])
b = np.array([2400, 100, -5000, 0])
A_Inv = sla.inv(matA)
M = np.dot(A_Inv, b)
M

array([14.75095785, 26.13793103, 36.75862069, 35.79310345])

In [18]:
M-res

array([ 6.65111889, 13.78596416, 19.86006997, 19.30034983])

Portanto o aumento na concentração de carbono foi de 13.73 mg/h

##### d)

In [19]:
A = np.array([[225, 0, -25, 0], [0, 155, 0, -105], [225, 0, -275, 50], [0, 5, 250, -255]])
b = np.array([1400, 100, -2000, 0])
resu = elim_gaussiana2(A, b, pivotamento=True)[2]
resu

array([ 8.10956329, 12.11410688, 16.98606965, 16.93034826])

In [20]:
resu-res

array([ 0.00972433, -0.23785999,  0.08751893,  0.43759464])

A instalação da rede reduziria a concentração de carbono na sala das crianças em 0.2376 mg/h

#### Exercico 8

In [25]:
A = np.array([[15,0.30,1],[17,0.4,1.2],[19,0.5,1.5]])
b = np.array([1,2,3])
elim_gaussiana2(A, b, pivotamento=True)[2]

array([-2.22222222e-01,  1.44444444e+01,  6.66133815e-15])

In [26]:
resultado = -2.22222222e-01*3.89 + 1.44444444e+01*0.095 + 6.66133815e-15*0.282
resultado

0.5077777744200019

#### Exercico 9

In [28]:
A = np.array([[0,55,-25],[0,-35,-4],[-25,-4,29]])
b = np.array([-200,-250,100])
elim_gaussiana2(A, b, pivotamento=True)[2]

array([18.15559809,  5.38277512, 19.84210526])